### PR TITLE
Fix compression and reduce code complexity

### PR DESCRIPTION
**Pull request by Zazama**
Compression was broken because the data length indicator was not passed to the frame body buffer but compression relies on it.

### CODE REVIEW

**Comment by pbricout** [comment]
Should you add a test for this case?

File: src/ID3Helpers.js [hunk 1]
```diff
@@ -128,20 +128,60 @@ function getFramesFromID3Body(ID3FrameBody, ID3Version, options = {}) {
             continue
         }
         const frameHeaderFlags = ID3Util.parseFrameHeaderFlags(bodyFrameHeader, ID3Version)
+        if(frameHeaderFlags.dataLengthIndicator) {
```
Comment (pbricout) [inline src/ID3Helpers.js:131]: I see `dataLenghtInicator` is only set for version 4, but apparently the version 3 supports it too according to the doc:
https://id3.org/id3v2.3.0#Frame_header_flags

https://github.com/Zazama/node-id3/blob/a6c784e8e0ecd3d4c32bae36042985199a229573/src/ID3Util.js#L191-L211

File: src/ID3Helpers.js [hunk 2]
```diff
@@ -128,20 +128,60 @@ function getFramesFromID3Body(ID3FrameBody, ID3Version, options = {}) {
...
+            bodyFrameSize -= 4
```
Comment (pbricout) [inline src/ID3Helpers.js:132]: I don't understand why this is `-4` isn't the header longer by 4 bytes when compressed?

File: src/ID3Helpers.js [hunk 3]
```diff
@@ -128,20 +128,60 @@ function getFramesFromID3Body(ID3FrameBody, ID3Version, options = {}) {
...
-        currentPosition += bodyFrameSize + textframeHeaderSize
-        frames.push({
+        const frame = {
             name: specName,
             flags: frameHeaderFlags,
             body: frameHeaderFlags.unsynchronisation ? ID3Util.processUnsynchronisedBuffer(bodyFrameBuffer) : bodyFrameBuffer
-        })
+        }
+        if(frameHeaderFlags.dataLengthIndicator) {
+            frame['dataLengthIndicator'] = ID3FrameBody.readInt32BE(currentPosition + textframeHeaderSize)
```
Comment (pbricout) [inline src/ID3Helpers.js:142]: Why not
```suggestion
            frame.dataLengthIndicator = ID3FrameBody.readInt32BE(currentPosition + textframeHeaderSize)
```

File: src/ID3Helpers.js (hunk 2 above)
Comment (Zazama) [reply src/ID3Helpers.js:132]: Yes, the header is longer by 4 bytes, but we have to take those 4 bytes away from the body, because normally, we assume that the body is of size (frame.length - 10) instead of (frame.length - 14).

The code is a bit confusing, I'll try to make it more understandable

File: src/ID3Helpers.js (hunk 1 above)
Comment (Zazama) [reply src/ID3Helpers.js:131]: Good catch, I will add it and add a v3 test

File: src/ID3Helpers.js (hunk 2 above)
Comment (Zazama) [reply src/ID3Helpers.js]: I've changed the code to make it a bit more clear and added comments but I chose not to add the data length indicator into the header because technically it is not part of the header

> This information is added after the frame header and before the frame data

**Comment by Zazama** [comment]
It might make sense to extract an ID3Frame into a real class because right now the code is spread out so much e.g.

```javascript
const frame = ID3Frame.fromBuffer(...)
```

then the class handles flags/indicators/unsynchronisation instead of the `getFramesFromID3Body` function
and instead of e.g. `ID3Frames[frameIdentifier].read(frame.body, ID3Version)`

```javascript
frame.getValue() // or something like that
```

File: index.js [hunk 4]
```diff
@@ -183,29 +183,29 @@ function update(tags, filebuffer, options, fn) {
 
     const updateFn = (currentTags) => {
         currentTags = currentTags.raw || {}
-        Object.keys(rawTags).map((specName) => {
-            const options = ID3Util.getSpecOptions(specName, 3)
+        Object.keys(rawTags).map((frameIdentifier) => {
```
Comment (pbricout) [inline index.js:186]: It is usually a good idea to separate the rename in a different commit, it makes it easier to review as we can skip the rename commit and focus on the other changes.

File: src/ID3Helpers.js [hunk 5]
```diff
@@ -96,55 +96,61 @@ module.exports.getTagsFromBuffer = function(filebuffer, options) {
...
-function getFramesFromID3Body(ID3FrameBody, ID3Version, options = {}) {
+function isFrameDiscardedByOptions(frameIdentifier, options) {
+    if(options.exclude instanceof Array && options.exclude.includes(frameIdentifier)) {
+        return true
+    }
+
+    return options.include instanceof Array && !options.include.includes(frameIdentifier)
+}
+
+function getFramesFromID3Body(ID3TagBody, ID3Version, options = {}) {
```
Comment (pbricout) [inline src/ID3Helpers.js:107]: I am still a bit confused with frame and tag, what the difference. My understanding is that the frame is the container containing the tag data. Correct?

File: src/ID3Helpers.js [hunk 6]
```diff
@@ -96,55 +96,61 @@ module.exports.getTagsFromBuffer = function(filebuffer, options) {
...
-        const bodyFrameBuffer = Buffer.alloc(bodyFrameSize)
-        ID3FrameBody.copy(bodyFrameBuffer, 0, currentPosition + textframeHeaderSize + (frameHeaderFlags.dataLengthIndicator ? 4 : 0))
+
+        const frameHeaderFlags = ID3Util.parseFrameHeaderFlags(frameHeader, ID3Version)
+        // Frames may have a 32-bit data length indicator appended after their header,
+        // if that is the case, the real body starts after those 4 bytes.
+        const frameBodyOffset = frameHeaderFlags.dataLengthIndicator ? 4 : 0
+        const bodyFrameBuffer = Buffer.alloc(frameBodySize - frameBodyOffset)
+        ID3TagBody.copy(bodyFrameBuffer, 0, currentPosition + frameHeaderSize + frameBodyOffset)
+
```
Comment (pbricout) [inline src/ID3Helpers.js:141]: Do we need a copy? If we do read only operations, we could use a view instead with `Buffer.slice`.
https://nodejs.org/docs/latest-v10.x/api/buffer.html#buffer_buf_slice_start_end
```js
const frameBody = ID3TagBody.slice(currentPosition + frameHeaderSize + frameBodyOffset)
```

File: src/ID3Helpers.js [hunk 7]
```diff
@@ -96,55 +96,61 @@ module.exports.getTagsFromBuffer = function(filebuffer, options) {
...
-    }
+    const frameIdentifierSize = (ID3Version === 2) ? 3 : 4
+    const frameHeaderSize = (ID3Version === 2) ? 6 : 10
 
-    while(currentPosition < ID3FrameBody.length && ID3FrameBody[currentPosition] !== 0x00) {
-        const bodyFrameHeader = Buffer.alloc(textframeHeaderSize)
-        ID3FrameBody.copy(bodyFrameHeader, 0, currentPosition)
+    while(currentPosition < ID3TagBody.length && ID3TagBody[currentPosition] !== 0x00) {
+        const frameHeader = Buffer.alloc(frameHeaderSize)
+        ID3TagBody.copy(frameHeader, 0, currentPosition)
```
Comment (pbricout) [inline src/ID3Helpers.js:119]: Do we need a copy? If we do read only operations, we could use a view instead with `Buffer.slice`.
https://nodejs.org/docs/latest-v10.x/api/buffer.html#buffer_buf_slice_start_end
```js
const frameHeader = ID3TagBody.slice(currentPosition)
```

**Review by pbricout — COMMENTED** [review]
Looks good some comments.

**Comment by pbricout** [comment]
> It might make sense to extract an ID3Frame into a real class because right now the code is spread out so much e.g.

That seems like a good idea, I am not super familiar with the code yet. That would handle a single frame or multiple frames?

```js
frame.getValue() // or something like that
```

What would `getValue` return exactly?

File: src/ID3Helpers.js (hunk 5 above)
Comment (Zazama) [reply src/ID3Helpers.js:107]: A frame is e.g. `TALB` (album), it contains information and is part of the ID3 tag.
An ID3 tag is the whole buffer containing the ID3 header and its frames.
Sometimes, a set of frames is called tags (not by the specs, but informally), which might be a bit confusing.

There are also parts in the code that are wrongly named from earlier times e.g. https://github.com/Zazama/node-id3/blob/master/src/ID3Util.js#L86 which is named `getFramePosition` even though it searches for the tag's position.

**Comment by Zazama** [comment]
Classes are supported, you can check here: https://node.green/
An `ID3Frame` would handle a single frame, for multiples it would be a tag (e.g. `ID3Tag`), which might also make sense to create.

`frame.getValue()` would be what the read functions in ID3Frames.js return, but without arrays, only single ones.
```javascript
// title string
let title = titleFrame.getValue();
```

and e.g. `frame.getBuffer()` would return the corresponding buffer which would be what the write functions return atm.

There would need to be a way to create a frame from a buffer and one to create it by value, e.g.

```javascript
const titleFrame = ID3Frame.createFromBuffer(buffer)
const title = titleFrame.getValue()

const albumFrame = ID3Frame.createFromValue('TALB', 'some album')
const albumFrameBuffer = albumFrame.getBuffer()
```

File: src/ID3Helpers.js (hunk 6 above)
Comment (Zazama) [reply src/ID3Helpers.js:141]: I don't think we need to copy but I wasn't sure if read-only is guaranteed throughout the whole code so I left it in for now

**Comment by Zazama** [comment]
I will try to create a new PR with an example

File: src/ID3Helpers.js (hunk 5 above)
Comment (pbricout) [reply src/ID3Helpers.js:107]: OK thanks for the explanation. We should fix same names in a follow up PR.

**Comment by pbricout** [comment]
> Classes are supported, you can check here: https://node.green/

Excellent, I was looking for something like that a bit earlier and could not find it! Really handy, now I know which exactly which features we can use.

> I will try to create a new PR with an example

Sounds good.

I think you can merge this one if you want. The buffer views would be nice but can be done if we better want to review the code.

**Comment by Zazama** [comment]
Ok I just checked quickly and the whole buffer in that function is a copy anyways, so it really doesn't matter if slice is used. I'll commit it later this week before merging

**Comment by pbricout** [comment]
> Ok I just checked quickly and the whole buffer in that function is a copy anyways, so it really doesn't matter if slice is used. I'll commit it later this week before merging

I would expect a function called `getFramesFromID3Body` (i.e. **get**) to not modify anything. We should strive to have that always true.

**Comment by pbricout** [comment]
> ```js
> const titleFrame = ID3Frame.createFromBuffer(buffer)
> const title = titleFrame.getValue()
> 
> const albumFrame = ID3Frame.createFromValue('TALB', 'some album')
> const albumFrameBuffer = albumFrame.getBuffer()
> ```

I see, it would know about the specific frames, I was thinking generic below. But why we can try to see what it looks like.

**Comment by Zazama** [comment]
Turns out, buffer.slice has been deprecated since v16. Instead, Buffer.subarray can be used and is basically the same API.
